### PR TITLE
Feat/arb sep chaido deploy

### PR DESCRIPTION
**Pull request by madhurMongia**
## PR-Codex overview
This PR focuses on enhancing the `VeaInboxArbToGnosis` contract and related components, including the addition of new functions and events, adjustments to epoch periods, and updates to package dependencies.

### Detailed summary
- Added `event AffirmationCompleted` in `IAMB.sol`.
- Introduced new functions in `validator-cli/src/utils/ethers.ts` for connecting to `RouterArbToGnosis` and `IAMB`.
- Updated epoch period from 2 hours to 1 hour in `01-arb-to-gnosis-outbox.ts`.
- Added new environment variables for Gnosis RPC in `.env.dist`.
- Introduced new functionality for challenge tracking in `watcherArbToGnosis.ts`.
- Added a new JSON deployment file for `VeaInboxArbToGnosisTestnet`.
- Updated `package.json` files to include `ts-node` in `devDependencies`.

> The following files were skipped due to too many changes: `contracts/deployments/arbitrumSepolia/VeaInboxArbToGnosisTestnet.json`, `contracts/deployments/sepolia/RouterArbToGnosisTestnet.json`, `contracts/deployments/chiado/VeaOutboxArbToGnosisTestnet.json`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

## Release Notes

- **New Features**
	- Enhanced deployment logic for Arbitrum to Gnosis communication.
	- Introduced new contracts and updated existing ones for improved message passing.
	- Added new provider functions for better interaction with blockchain contracts.
	- New event `AffirmationCompleted` added to the Gnosis interface for tracking message affirmation statuses.

- **Bug Fixes**
	- Improved error handling and retry mechanisms for claims and challenges.

- **Documentation**
	- Updated deployment details and metadata for new and existing contracts.

- **Chores**
	- Updated configuration variables to reflect changes in supported networks.

### CODE REVIEW

File: contracts/deploy/01-outbox/01-arb-to-gnosis-outbox.ts [hunk 1]
```diff
@@ -84,27 +89,50 @@ const deployOutbox: DeployFunction = async (hre: HardhatRuntimeEnvironment) => {
...
-    nonce += 4; // SenderGatewayToEthereum deploy tx will be the 5th after this, same network for both sender/receiver.
 
-    const routerAddress = getContractAddress(deployer, nonce);
-    console.log("calculated future router for nonce %d: %s", nonce, routerAddress);
+    const routerAddress = getContractAddress(deployer, nonce + 10);
+    console.log("calculated future router for nonce %d: %s", nonce + 10, routerAddress);
+
+    const senderGatewayAddress = getContractAddress(deployer, nonce + 6); // with the current order of transaction ,nonce for sender gateway would be 14.
+    console.log("calculated future SenderGatewayToGnosis address for nonce %d: %s", nonce + 6, senderGatewayAddress);
+
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply contracts/deploy/01-outbox/01-arb-to-gnosis-outbox.ts:98]: getContractAddress() is the part of legacy code, its fine as the current goal is to resurrect the testnets, can use create2 when continuing the development.

File: contracts/deploy/02-inbox/02-arb-to-gnosis-inbox.ts [hunk 2]
```diff
@@ -37,20 +44,64 @@
...
+    });
+
+    await deploy("ArbToGnosisSenderGateway", {
+      from: deployer,
+      contract: "SenderGatewayMock",
+      args: [veaInbox.address, receiverGateway.address],
+      gasLimit: 4000000,
+      log: true,
+    });
+  };
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply contracts/deploy/02-inbox/02-arb-to-gnosis-inbox.ts:81]: already answered

File: validator-cli/src/utils/arbMsgExecutor.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    if (!childToParentMessage) {
+      throw new Error("No child-to-parent messages found");
+    }
+  } catch (error) {
+    throw new Error(`Failed to retrieve child-to-parent messages: ${error.message}`);
+  }
+
+  // Execute the message
+  try {
+    const res = await childToParentMessage.execute(childProvider);
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/utils/arbMsgExecutor.ts:44]: childToParentMessage is already aware of parent (eth) RPC

File: validator-cli/src/utils/arbMsgExecutor.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import { Signer } from "@ethersproject/abstract-signer";
+import { ContractTransaction } from "@ethersproject/contracts";
+
+// Execute the child-to-parent (arbitrum-to-ethereum) message, for reference see: https://docs.arbitrum.io/sdk/reference/message/ChildToParentMessage
+export default async function messageExecutor(
+  trnxHash: string,
+  childRpc: string,
+  parentRpc: string
+): Promise<ContractTransaction> {
+  const PRIVATE_KEY = process.env.PRIVATE_KEY;
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/utils/arbMsgExecutor.ts:13]: Wallet class already has private key validation

File: relayer-cli/src/devnetRelayExample.ts [hunk 5]
```diff
@@ -2,8 +2,8 @@ import { relayAllFrom } from "./utils/relay";
 import * as fs from "fs";
 
 // let chain_ids = [5, 10200];
-let chain_ids = [11155111];
-const epochPeriod = 1800; // 30 min
+let chain_ids = [10200, 11155111];
+const epochPeriod = 3600; // 1 hour
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply relayer-cli/src/devnetRelayExample.ts:6]: above files are not part of the current scope

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 6]
```diff
@@ -243,232 +306,90 @@ const watch = async () => {
...
-                if (!txnReceipt) {
-                  console.log("challenge txn " + challengeTxnHashes.get(index) + " not mined yet");
-                  continue;
-                }
-                const blockNumber = txnReceipt.blockNumber;
-                const challengeBlock = (await retryOperation(
-                  () => providerGnosis.getBlock(blockNumber),
+            )) as ContractTransaction;
+            console.log("Epoch " + veaEpochOutboxCheck + " challenged with txn " + txnChallenge.hash);
+            continue;
```
Comment (mani99brar) [inline validator-cli/src/ArbToEth/watcherArbToGnosis.ts:350]: Before `continue` the `challengeProgress.challenge.status` should be set to "pending"?

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 7]
```diff
@@ -243,232 +306,90 @@ const watch = async () => {
...
-                  )) as ContractTransaction;
-
-                  challengeTxnHashes.set(index, txnChallenge.hash);
-                  console.log(
-                    "challenging claim for epoch " + veaEpochOutboxCheck + " with txn hash " + txnChallenge.hash
-                  );
-                }
+                )) as ContractTransaction;
+                console.log("Epoch " + veaEpochOutboxCheck + " sendSnapshot called with txn " + txnSendSnapshot.hash);
+              }
```
Comment (mani99brar) [inline validator-cli/src/ArbToEth/watcherArbToGnosis.ts:362]: I think `challengeProgress.snapshot.status` should be updated to "pending" with the sendSnapshot txHash?

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 8]
```diff
@@ -682,7 +604,499 @@ const ArbBlockToL1Block = async (
...
+    const arbCost = arbGasEstimate.mul(arbGasPrice);
+
+    if (arbCost.gt(maximumAllowableCost)) {
+      return { profitable: false, estimatedProfit: constants.Zero };
+    }
+    totalCost = totalCost.add(arbCost);
+    maximumAllowableCost = maximumAllowableCost.sub(arbCost);
+
+    // 3. Costs on Ethereum (for Arbitrum -> Ethereum message)
+    //TODO : L2 to L1 message execution gas cost
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:804]: estimateL1Gas method doesn't exist on childToParentMessage

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 9]
```diff
@@ -682,7 +622,512 @@ const ArbBlockToL1Block = async (
...
+        ...challengeProgress,
+        status: "WithdrawalPending",
+      };
+    }
+  }
+
+  // there is no event in case of withdrawal hence no way to track it ,
+  // but if a withdrawal is processed ,claimHash for the epoch will be deleted ,challenged progess will not be recontructed in the first place.
+  return challengeProgress;
+}
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:1112]: unnecessary abstraction

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts (hunk 6 above)
Comment (mani99brar) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:361]: Looks fixed 👍

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts (hunk 7 above)
Comment (mani99brar) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:381]: Looks fixed 👍

**Review by mani99brar — APPROVED** [review]

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 10]
```diff
@@ -3,23 +3,79 @@ import {
...
+import { Block, Log } from "@ethersproject/abstract-provider";
 import { SequencerInbox } from "@arbitrum/sdk/dist/lib/abi/SequencerInbox";
+import { NodeInterface } from "@arbitrum/sdk/dist/lib/abi/NodeInterface";
+import {
+  IAMB,
+  RouterArbToGnosis,
+  VeaInboxArbToGnosis,
+  VeaOutboxArbToGnosis,
+} from "@kleros/vea-contracts/typechain-types";
+import { ClaimStruct } from "@kleros/vea-contracts/typechain-types/arbitrumToEth/VeaInboxArbToEth";
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:24]: both uses the same claim struct

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 11]
```diff
@@ -243,232 +305,90 @@
...
+            blockNumberOutboxLowerBound,
+            amb,
+            previousProgress
+          );
+          challenges.set(index, challengeProgress);
+          console.log(
+            "challenge progess for epoch " + veaEpochOutboxCheck + "  is " + JSON.stringify(challengeProgress)
+          );
+          //TODO : check profitablity of the whole dispute resolution
+          //const profitablity = await calculateDisputeResolutionProfitability(veaEpochOutboxCheck,claim,veaOutbox,veaInbox,providerGnosis,providerArb,providerEth);
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:344]: not important for resurrecting the testnet ,will be picked up when the developement starts

[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:344]: not right now

**Review by mani99brar — APPROVED** [review]

**Review by jaybuidl — DISMISSED** [review]

File: validator-cli/src/utils/ethers.ts [hunk 12]
```diff
@@ -63,6 +65,13 @@ function getVeaOutboxArbToGnosis(veaOutboxAddress: string, privateKey: string, w
   return VeaOutboxArbToGnosisDevnet__factory.connect(veaOutboxAddress, getWallet(privateKey, web3ProviderURL));
 }
 
+function getVeaRouterArbToGnosisProvider(veaRouterAddress: string, privateKey: string, rpc: JsonRpcProvider) {
+  return RouterArbToGnosis__factory.connect(veaRouterAddress, getWalletRPC(privateKey, rpc));
+}
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/utils/ethers.ts:70]: .connect already has input parameter validation

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 13]
```diff
@@ -682,7 +622,511 @@ const ArbBlockToL1Block = async (
...
+      claim.honest,
+      claim.challenger,
+    ]
+  );
+};
+
 (async () => {
-  await watch();
+  retryOperation(() => watch(), 1000, 10);
 })();
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:1131]: currently the bot only has TCP connections to be closed, both client and server are already resilient when it comes to TCP  connections

File: validator-cli/src/ArbToEth/watcherArbToGnosis.ts [hunk 14]
```diff
@@ -682,7 +622,511 @@ const ArbBlockToL1Block = async (
...
+  veaOutbox: VeaOutboxArbToGnosis,
+  veaInbox: VeaInboxArbToGnosis,
+  router: RouterArbToGnosis,
+  providerGnosis: JsonRpcProvider,
+  providerArb: JsonRpcProvider,
+  providerEth: JsonRpcProvider,
+  blockNumberOutboxLowerBound: number,
+  amb: IAMB,
+  previousProgress?: ChallengeProgress
+): Promise<ChallengeProgress> {
```
[the remark this answers is not in the document]
Comment (madhurMongia) [reply validator-cli/src/ArbToEth/watcherArbToGnosis.ts:840]: redundant error handling

**Review by mani99brar — APPROVED** [review]

**Pull request merged**